### PR TITLE
[MOSTLY MODULAR] Oppressive Force Relocation Medicell

### PR DESCRIPTION
**Pull request by softcerv**
## About The Pull Request
Adds in a new Medicell, the Oppressive Force Relocation Medicell. When shot at someone inside surgery, aux surgery, treatment center, or medical storage, without surgery access, they will be given 10 seconds to leave the area before being teleported back to the medbay lobby.

There is an admin version, `/obj/item/weaponcell/medical/utility/relocation/upgraded` that lacks the delay, works anywhere on the station, and can teleport anyone no matter the access on their ID.

The teleport destination can be changed through var-editing. (You need the ammo_type in the gun, after the cell has been inserted.)

## How This Contributes To The Skyrat Roleplay Experience
This gives medical staff a non-violent way to discourage people from crowding in sensitive areas. By giving doctors a way to peacefully remove security and other invaders from sensitive areas, medical doctors can focus more on doing their job and roleplaying with patients.
 

## Changelog

:cl:
add: Adds in the Oppressive Force Relocation Medicell. It can be used to teleport non-medical staff out of sensitive areas.
/:cl:

### CODE REVIEW

File: modular_skyrat/modules/modular_weapons/code/medcells.dm [hunk 1]
```diff
@@ -557,6 +557,111 @@
...
+	teleport_areas = list(/area/medical/surgery, /area/medical/treatment_center, /area/medical/storage, /area/medical/patients_rooms)
+	grace_period = TRUE
+	access_teleporting = TRUE
+
+/obj/projectile/energy/medical/utility/relocation/on_hit(mob/living/target)
+	. = ..()
+
+	var/mob/living/carbon/human/teleportee = target
+
+	if(area_locked && teleport_areas.len && !is_type_in_list(get_area(target), teleport_areas))
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_weapons/code/medcells.dm]: use length() not .len

File: modular_skyrat/modules/modular_weapons/code/medcells.dm [hunk 2]
```diff
@@ -557,6 +557,111 @@
...
+		for(var/turf/turf_in_area in get_area_turfs(destination_area))
+			if(!turf_in_area.is_blocked_turf())
+				turf_list += turf_in_area
+
+	teleportee.visible_message(span_notice("[teleportee] is teleported away!"))
+
+	do_teleport(teleportee, pick(turf_list), no_effects = FALSE, channel = TELEPORT_CHANNEL_QUANTUM)
+
+
+/datum/component/medigun_relocation
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_weapons/code/medcells.dm:630]: code doc comments

File: modular_skyrat/modules/modular_weapons/code/medcells.dm [hunk 3]
```diff
@@ -557,6 +557,111 @@
...
+		var/target_access = teleportee.wear_id.GetAccess() //Stores the access of the target within a variable
+		if(required_access in target_access)
+			return FALSE
+
+	if(grace_period)
+		to_chat(teleportee, span_warning("You have [time_allowance] to leave, if you do not leave in this time, you will be forcibly teleported outside."))
+		teleportee.AddComponent(/datum/component/medigun_relocation, time_allowance, destination_area, area_locked, teleport_areas)
+		return TRUE
+
+	if(area_locked && teleport_areas.len && !is_type_in_list(get_area(teleportee), teleport_areas)) // Does one final check to see if the target is in the area.
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_weapons/code/medcells.dm]: length over len

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: modular_skyrat/modules/modular_weapons/code/medcells.dm [hunk 4]
```diff
@@ -557,6 +557,108 @@
...
+			if(!turf_in_area.is_blocked_turf())
+				turf_list += turf_in_area
+
+	teleportee.visible_message(span_notice("[teleportee] is teleported away!"))
+
+	do_teleport(teleportee, pick(turf_list), no_effects = FALSE, channel = TELEPORT_CHANNEL_QUANTUM)
+
+/// Used to handle teleporting if there is a grace period
+/datum/component/medigun_relocation
+	var/area/destination_area
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_weapons/code/medcells.dm:632]: Code doc the vars, I meant.

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: modular_skyrat/modules/modular_weapons/code/medcells.dm (hunk 4 above)
Comment (softcerv) [reply modular_skyrat/modules/modular_weapons/code/medcells.dm:632]: Gotcha

**Review by Zonespace27 — APPROVED** [review]

**Pull request merged**